katsdptelstate/telescope_state.py

```diff
@@ -361,9 +400,90 @@ def add(self, key, value, ts=None, immutable=False, encoding=ENCODING_DEFAULT):
             try:
                 self._backend.add_mutable(full_key, str_val, ts)
             except ImmutableKeyError:
-                raise ImmutableKeyError('Attempt to overwrite immutable key {}'
+                raise ImmutableKeyError('Attempt to change key {} to mutable'
                                         .format(key_str))
 
+    def set_indexed(self, key, sub_key, value, encoding=ENCODING_DEFAULT):
+        """Set a sub-key of an indexed key.
+
+        Parameters
+        ----------
+        key : str or bytes
+            Main key
+        sub_key : object
+            Sub-key within `key` to associate with the value. It must be both
+            hashable and serialisable.
+        encoding : bytes
+            Encoding used for `value` (see :func:`encode_value`). Note that it
+            does not affect the encoding of `sub_key`.
+
+        Raises
+        ------
+        InvalidKeyError
+            if `key` collides with a class member name
+        ImmutableKeyError
+            if the sub-key already exists with a different value
+        ImmutableKeyError
+            if the key already exists and is not indexed
+        redis.ResponseError
+            if there is some other error from the Redis server
+        """
+        # check that we are not going to munge a class method
+        if key in self.__class__.__dict__:
+            raise InvalidKeyError("The specified key already exists as a "
+                                  "class method and thus cannot be used.")
+        # Raises a TypeError if it's not hashable, to prevent trouble
+        # retrieving it later.
+        hash(sub_key)
+        key = ensure_binary(key)
+        full_key = self._prefixes[0] + key
+        key_str = display_str(full_key)
+        # Sub-keys will always be encoded with ENCODING_MSGPACK, so that
+        # lookups don't need to worry whether they are using the matching
+        # encoding.
+        sub_key_enc = encode_value(sub_key, ENCODING_MSGPACK)
```

Comment:
If someone overrides ENCODING_DEFAULT don't we want to change the sub_key encoding to match ?

Reply:
No, because it has to be consistent across all users of a shared telstate for `get_indexed` to work. For example, suppose we hypothetically decided that CBOR is better than msgpack and changed telstate so that ENCODING_DEFAULT was CBOR. If we used ENCODING_DEFAULT here, then the archive would end up with some .rdb files where subkeys are encoded with msgpack and others with CBOR. In theory you could even end up with the same subkey appearing twice, once with each encoding. That would make `get_indexed` much more complicated. I'm not expecting sub-keys to have particularly complicated types so I don't think committing to msgpack here is going to be a major impediment.